### PR TITLE
Braulio/03 lazy loading

### PR DESCRIPTION
**Pull request by brauliodiez**
Lazy loading POC:

- Integrated with react router.
- Using related child components.
- Using require.ensure
- Two bundles created.

About the sample: we assume the students page is only loaded on demand, whenever the react-router get's a navigation order to that page it will load it asynchronously (and related child components will be loaded as well).

### CODE REVIEW

File: frontend/src/index.tsx [hunk 1]
```diff
@@ -2,16 +2,39 @@ import * as React from 'react';
...
+
 
 import { Router, Route, IndexRoute, Link, IndexLink, browserHistory, hashHistory  } from 'react-router'
 
+
+// If it needs async loading it will just call require.ensure (webpack)
+const component = (component : any) : any => {
+  // TODO:Removed ... and {} in return check why this is needed
+  // https://medium.com/@Nadav.Dav/implementing-lazy-loading-with-webpack-and-react-router-f8497f895892#.mq4ug0c15
+  const isReactComponent = (obj) => Boolean(obj && obj.prototype && Boolean(obj.prototype.isReactComponent));
```
Comment (JaimeSalas) [inline frontend/src/index.tsx]: If you are using TypeScript (sorry I can not figure out with the code), maybe you can use instanceof for isReactComponent. Any way smells like something that could be promote to a library.

Comment (JaimeSalas) [inline frontend/src/index.tsx]: Sorry I watch the requires bellow,  TypeScript at last.

File: frontend/src/index.tsx [hunk 2]
```diff
@@ -2,16 +2,39 @@ import * as React from 'react';
...
+
+  if(isReactComponent(component)) {
+    return component;
+  } else {
+    return {
+      getComponent: (location, cb) => {
+        (require as any).ensure([], require => {
```
Comment (JaimeSalas) [inline frontend/src/index.tsx]: cb?

File: frontend/src/index.tsx (hunk 1 above)
Comment (brauliodiez) [inline frontend/src/index.tsx]: You are right, this code can be improved. This function could be use just to use always the same approach (do not differentiate whether is a component or async request), but not sure if it's worth to go that path.

File: frontend/src/index.tsx (hunk 2 above)
Comment (brauliodiez) [inline frontend/src/index.tsx]: I think it's an achronym coming from  "callback"

File: frontend/src/index.tsx (hunk 1 above)
Comment (JaimeSalas) [inline frontend/src/index.tsx]: Enough right now. The effort has to be applied only when it is worth enough, not the case.

File: frontend/src/index.tsx [hunk 3]
```diff
@@ -2,17 +2,39 @@ import * as React from 'react';
...
+const component = (component : any) : any => {
+  // https://medium.com/@Nadav.Dav/implementing-lazy-loading-with-webpack-and-react-router-f8497f895892#.mq4ug0c15
+  const isReactComponent = (obj) => Boolean(obj && obj.prototype && Boolean(obj.prototype.isReactComponent));
+
+  if(isReactComponent(component)) {
+    return component;
+  } else {
+    // Async chunk loading
+    return {
+      getComponent: (location, cb) => {
```
Comment (Nasdan) [inline frontend/src/index.tsx]: cb is callback?

File: frontend/typings.json [hunk 4]
```diff
@@ -2,6 +2,7 @@
   "name": "lm-app",
   "dependencies": {},
   "globalDependencies": {
+    "node": "registry:dt/node#6.0.0+20160720070758",
```
Comment (Nasdan) [inline frontend/typings.json]: Why do yo need node?

If you want "require" intellisense and avoid `(require as any)` you can use [webpack-env](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/webpack/webpack-env.d.ts) typings

File: frontend/src/index.tsx (hunk 3 above)
Comment (Nasdan) [inline frontend/src/index.tsx]: Yes, Now I see your comments

File: frontend/src/index.tsx [hunk 5]
```diff
@@ -2,17 +2,39 @@ import * as React from 'react';
...
-import {Students} from './components/pages/students'
+
 
 import { Router, Route, IndexRoute, Link, IndexLink, browserHistory, hashHistory  } from 'react-router'
 
+
+// If it needs async loading it will just call require.ensure (webpack)
+const component = (component : any) : any => {
+  // https://medium.com/@Nadav.Dav/implementing-lazy-loading-with-webpack-and-react-router-f8497f895892#.mq4ug0c15
+  const isReactComponent = (obj) => Boolean(obj && obj.prototype && Boolean(obj.prototype.isReactComponent));
```
Comment (Nasdan) [inline frontend/src/index.tsx]: [Boolean function](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Boolean) is from Javascript, right? I've never used it

File: frontend/typings.json (hunk 4 above)
Comment (brauliodiez) [inline frontend/typings.json]: Really good point, please could you add it in this branch and refactor? Thanks

File: frontend/src/index.tsx (hunk 5 above)
Comment (brauliodiez) [inline frontend/src/index.tsx]: Not sure, we have to take a look :-)

File: frontend/package.json [hunk 6]
```diff
@@ -24,14 +24,14 @@
   },
   "homepage": "https://github.com/MasterLemon2016/LeanMood#readme",
   "devDependencies": {
+    "html-webpack-plugin": "^2.22.0",
```
Comment (Nasdan) [inline frontend/package.json]: I move this to devDependencies

File: frontend/src/index.tsx [hunk 7]
```diff
@@ -2,17 +2,39 @@ import * as React from 'react';
...
+
+  if(isReactComponent(component)) {
+    return component;
+  } else {
+    // Async chunk loading
+    return {
+      getComponent: (location, cb) => {
+        require.ensure([], require => {
+              // Retrieve product page component
+              cb(null, require('./components/pages/students')["default"]);
```
Comment (Nasdan) [inline frontend/src/index.tsx]: default property does not have intellisense with webpack-env. So, I had to write like `["default"]`

**Pull request merged**